test/unit/callbacks_and_delays_test.lua

```diff
@@ -0,0 +1,127 @@
+local expirationd = require("expirationd")
+local fiber = require("fiber")
+local t = require("luatest")
+
+local helpers = require("test.helper")
+
+local g = t.group('callbacks_and_delays', {
+    {index_type = 'TREE', engine = 'vinyl'},
+    {index_type = 'TREE', engine = 'memtx'},
+    {index_type = 'HASH', engine = 'memtx'},
+})
+
+g.before_each({index_type = 'TREE'}, function(cg)
+    t.skip_if(cg.params.engine == 'vinyl' and helpers.vinyl_is_broken(),
+            'Blocked by https://github.com/tarantool/tarantool/issues/6448')
+    g.space = helpers.create_space_with_tree_index(cg.params.engine)
+end)
+
+g.before_each({index_type = 'HASH'}, function(cg)
+    g.space = helpers.create_space_with_hash_index(cg.params.engine)
+end)
+
+g.before_each({index_type = 'BITSET'}, function(cg)
+    g.space = helpers.create_space_with_bitset_index(cg.params.engine)
+end)
+
+g.after_each(function(g)
+    g.space:drop()
+end)
+
+function g.test_delays_and_scan_callbacks(cg)
```

Comment:
```
sergeyb@pony:~/sources/MRG/expirationd$ ./.rocks/bin/luatest -v -c 'callbacks_and_delays.engine:"vinyl".index_type:"TREE".test_delays_and_scan_callbacks'
...
2021-11-29 14:22:57.290 [1031066] main/107/vinyl.scheduler I> dumped 49184 bytes in 0.0 s, rate 10.2 MB/s
2021-11-29 14:22:58.095 [1031066] main/118/guardian of "test" I> expiration: task "test" restarted
2021-11-29 14:22:58.096 [1031066] main/121/worker of "test" utils.c:449 E> LuajitError: table: 0x4144f1f8
2021-11-29 14:22:59.095 [1031066] main/118/guardian of "test" I> expiration: task "test" restarted
2021-11-29 14:22:59.096 [1031066] main/122/worker of "test" utils.c:449 E> LuajitError: table: 0x40ccf8a0
2021-11-29 14:23:00.096 [1031066] main/118/guardian of "test" I> expiration: task "test" restarted
2021-11-29 14:23:00.097 [1031066] main/123/worker of "test" utils.c:449 E> LuajitError: table: 0x40cd0578
2021-11-29 14:23:01.097 [1031066] main/118/guardian of "test" I> expiration: task "test" restarted
2021-11-29 14:23:01.098 [1031066] main/124/worker of "test" utils.c:449 E> LuajitError: table: 0x40cd1000
2021-11-29 14:23:02.097 [1031066] main/118/guardian of "test" I> expiration: task "test" restarted
2021-11-29 14:23:02.098 [1031066] main/125/worker of "test" utils.c:449 E> LuajitError: table: 0x408b0380
2021-11-29 14:23:03.098 [1031066] main/118/guardian of "test" I> expiration: task "test" restarted
2021-11-29 14:23:03.099 [1031066] main/126/worker of "test" utils.c:449 E> LuajitError: table: 0x40cd2670
2021-11-29 14:23:04.099 [1031066] main/118/guardian of "test" I> expiration: task "test" restarted
2021-11-29 14:23:04.100 [1031066] main/127/worker of "test" utils.c:449 E> LuajitError: table: 0x40cd30e8
^C2021-11-29 14:23:04.365 [1031066] main I> tx_binary: stopped
```

Reply:
It seems that something is wrong here, in the comments it says `< 2.8.2`, and in the code` <= 2.8.2`

https://github.com/tarantool/expirationd/blob/838c2d16346741cd753f6a95810284ecb4c6b50a/test/helper.lua#L258-L259

And please tell me which version you are using, because I could not find tarantool for macos on s3 with version > 2.8.2, I don't really want to do it from under docker

Reply:
The comment and code are consistent. Bug https://github.com/tarantool/tarantool/issues/6448 fixed in versions >= 2.8.2. For versions without fix, we skip tests on vinyl engine.

Reply:
I am testing on 2.8.2 and the test is skipping

```
./.rocks/bin/luatest -v -c 'callbacks_and_delays.engine:"vinyl".index_type:"TREE".test_delays_and_scan_callbacks'
Started on Mon Nov 29 16:43:06 2021
2021-11-29 16:43:06.303 [3059] main/103/luatest C> Tarantool Enterprise 2.8.2-0-gfc96d10f5-r427-macos
2021-11-29 16:43:06.303 [3059] main/103/luatest C> log level 5
2021-11-29 16:43:06.303 [3059] main/103/luatest I> wal/engine cleanup is paused
2021-11-29 16:43:06.303 [3059] main/103/luatest I> mapping 268435456 bytes for memtx tuple arena...
2021-11-29 16:43:06.303 [3059] main/103/luatest I> Actual slab_alloc_factor calculated on the basis of desired slab_alloc_factor = 1.044274
2021-11-29 16:43:06.304 [3059] main/103/luatest I> mapping 16777216 bytes for vinyl tuple arena...
2021-11-29 16:43:06.308 [3059] main/103/luatest I> instance uuid a834541a-2b78-4716-b82d-634d1584358c
2021-11-29 16:43:06.308 [3059] main/103/luatest I> tx_binary: stopped
2021-11-29 16:43:06.308 [3059] main/103/luatest I> initializing an empty data directory
2021-11-29 16:43:06.327 [3059] main/103/luatest I> assigned id 1 to replica a834541a-2b78-4716-b82d-634d1584358c
2021-11-29 16:43:06.327 [3059] main/103/luatest I> cluster uuid 4953dc5e-af3f-4d9a-936d-9f6f63c1ce90
2021-11-29 16:43:06.328 [3059] snapshot/101/main I> saving snapshot `/var/folders/yp/ymwg4wxs381bmvgjp70khxy40000gp/T//AqMdVw/00000000000000000000.snap.inprogress'
2021-11-29 16:43:06.329 [3059] snapshot/101/main I> done
2021-11-29 16:43:06.330 [3059] main/103/luatest I> ready to accept requests
2021-11-29 16:43:06.330 [3059] main/104/gc I> wal/engine cleanup is resumed
2021-11-29 16:43:06.330 [3059] main/103/luatest I> set 'log_level' configuration option to 5
2021-11-29 16:43:06.330 [3059] main/105/checkpoint_daemon I> scheduled next checkpoint for Mon Nov 29 17:43:57 2021
2021-11-29 16:43:06.330 [3059] main/103/luatest I> set 'vinyl_memory' configuration option to 1024
2021-11-29 16:43:06.331 [3059] main/103/luatest I> set 'log_format' configuration option to "plain"
    callbacks_and_delays.engine:"vinyl".index_type:"TREE".test_delays_and_scan_callbacks ...
    callbacks_and_delays.engine:"vinyl".index_type:"TREE".test_delays_and_scan_callbacks ... (0.000s) skip
...tool/expirationd/test/unit/callbacks_and_delays_test.lua:14: Blocked by https://github.com/tarantool/tarantool/issues/6448
=========================================================
Ran 0 tests in 0.030 seconds, 0 successes, 0 failures, 1 skipped
2021-11-29 16:43:06.335 [3059] main I> tx_binary: stopped
```

Reply:
Also in CI, there is 2.8.2 and it also skips these tests

Reply:
I'm testing on Tarantool 2.10.0-beta1-0-g7da4b1438 and see errors:
```
2021-12-07 10:12:00.931 [2556718] main/118/guardian of "test" I> expiration: task "test" restarted
2021-12-07 10:12:00.932 [2556718] main/121/worker of "test" utils.c:449 E> LuajitError: table: 0x405deb98
2021-12-07 10:12:01.932 [2556718] main/118/guardian of "test" I> expiration: task "test" restarted
2021-12-07 10:12:01.933 [2556718] main/122/worker of "test" utils.c:449 E> LuajitError: table: 0x405df630
2021-12-07 10:12:02.932 [2556718] main/118/guardian of "test" I> expiration: task "test" restarted
2021-12-07 10:12:02.933 [2556718] main/123/worker of "test" utils.c:449 E> LuajitError: table: 0x405e00b8
```

How it can be related to https://github.com/tarantool/tarantool/issues/6448?

Reply:
```diff
--- a/test/unit/callbacks_and_delays_test.lua
+++ b/test/unit/callbacks_and_delays_test.lua
@@ -45,7 +45,6 @@ function g.test_delays_and_scan_callbacks(cg)
     local iteration_delay = 1
     local full_scan_delay = 2
     local full_scan_success_counter = 0
-
     local task = expirationd.start("test", space.id, helpers.is_expired_true,
             {
                 iteration_delay = iteration_delay,
@@ -57,7 +56,7 @@ function g.test_delays_and_scan_callbacks(cg)
                         -- Check the full scan delay with an accuracy of 0.1 seconds.
                         -- Difference between start time of the second full scan
                         -- and complete_time of the first full scan.
-                        t.assert_almost_equals(start_time - complete_time, full_scan_delay, 0.1)
+                        --t.assert_almost_equals(start_time - complete_time, full_scan_delay, 0.1)
                     end
                 end,
                 on_full_scan_success = function(_)
@@ -71,7 +70,7 @@ function g.test_delays_and_scan_callbacks(cg)
                     else
                         -- Check the accuracy of iteration delay.
                         -- Difference between start time and complete_time of the first full scan.
-                        t.assert_almost_equals(complete_time - start_time, iteration_delay, 0.1)
+                        --t.assert_almost_equals(complete_time - start_time, iteration_delay, 0.1)
                         first_fullscan_done = true
                     end
                 end
```
Error in luatest assert. If we remove these checks, then there will be no errors

Reply:
It is generally a problem that `t.assert` inside a fiber works like this. It will not throw an error in `luatest` and we will not know if the assert has failed.

Reply:
Fixed. After all the fixes, I did not get any errors on Tarantool 2.10